### ISSUE TITLE
django.http.request.RawPostDataException: You cannot access body after reading from request's data stream

### ISSUE THREAD

**Issue by pquentin** [tomchristie/django-rest-framework#2774]
When sending a POST request through django-rest-framework, I get this (reformatted) exception:

    django.http.request.RawPostDataException:
    You cannot access body after reading from request's data stream

What's going on?

**Comment by pquentin** [comment]
(I spent a few hours on this and I'm [answering to myself StackOverflow-style](http://stackoverflow.com/help/self-answer) in order to help others and raise awareness about the issue to the awesome django-rest-framework developers.)

For various reasons, Django only allows reading POST body **once**. Since the body is exposed with a file-like API, you can't easily read the file again. Also, you won't be able to [change upload file handlers](https://docs.djangoproject.com/en/dev/topics/http/file-uploads/#id1) if you have already read the data. Failing to read POST data only once will fail, and it is tricky to debug for two reasons.

 1. Django only reports its error the second time you access the body, so it can be difficult to track the first time it was accessed.
 2. A myriad of different issues may cause this. One way is to print the full traceback (with `traceback.print_stack()`) whenever the data gets read (in `django/http/request.py`: `HttpRequest.read` and `HttpRequest.readline`). This is how I did it, maybe there are other ways.

So, did you access `request.method` somewhere in your code? Are you using Django test client? If so, it's possible that [HTTP header based method overriding](http://www.django-rest-framework.org/topics/browser-enhancements/#http-header-based-method-overriding) kicked in. This feature allows browsers to simulate requests other than GET/POST. To do so, django-rest-framework looks at hidden form fields such as `<input type="hidden" name="_method" value="DELETE">`. With POST requests, this information is in the request body, so django-rest-framework will have to read the request body.

This feature is [enabled by default](http://www.django-rest-framework.org/api-guide/settings/#browser-overrides), but django-rest-framework ensures that the request is indeed POST and is indeed using a Content-Type that would be sent on a form submission. But that's exactly the behavior of [Django test client](https://docs.djangoproject.com/en/dev/topics/testing/tools/#the-test-client)! Two possible fixes exist:

1. Disable browser overrides

        REST_FRAMEWORK = {
            'FORM_METHOD_OVERRIDE': None,
            'FORM_CONTENT_OVERRIDE': None,
            'FORM_CONTENTTYPE_OVERRIDE': None
        }

2. Change the Content-Type in Django test client

        from django.test import Client

        client = Client()
        response = client.post(url, content_type='application/json)

**Comment by pquentin** [comment]
Incredible, that was exactly it, looks like you read in my mind (and code)! :)

**Comment by tomchristie** [comment]
Favorite kind of support issue. Self closing, helps others out. Thanks! :sparkles:

**Comment by naveenrao18** [comment]
I am trying to return a HttpResponse from another view (called from django-rest-framework view method) & I still run into this same error. The above mentioned fixes do not seem to work :(
Note: This happens when I try to read request.POST data in the view method that's returning HttpResponse object.

**Comment by james-aiwip** [comment]
I'm still getting this issue, it seems that the rest-framework reads the request in the request factory

**Comment by BHushanRathod** [comment]
i'm getting this error whenever i'm trying to call Class based view from a generic view. I have conditions in generic view based on which particular class view is called. In Class view whenever i try to access `request.data`, i get _django.http.request.RawPostDataException_ error. Any help?

```
@api_view(['GET', 'POST'])
def HangoutPlaceSport(request):
    choice = request.data['sport']
    if choice == 'Hangout_sport':
    return hangout_sport.as_view()(request)

class hangout_sport(APIView):
    def post(self,request)
    data = request.data
```

**Comment by Rydez** [comment]
I'm also getting this error and can't determine what's causing it. I'm trying to use ajax to perform a post request. 

```
$.ajax({
      type: 'POST',
      url: 'http://127.0.0.1:8000/api/accounts/register/',
      data: JSON.stringify({something: 4}),
      cache: false,
      contentType: 'application/json',
      dataType: 'json',
      success: function(data) {
        console.log(data);
      },
      error: function(err) {
        console.log(err);
      }
    })
    return false;
  });

```
```
class UserCreateAPIView(CreateAPIView):
  permission_classes = [AllowAny]
  serializer_class = UserCreateSerializer

  def post(self, request, *args, **kwargs):
    serializer = UserCreateSerializer(data=request.data)
```

`django.http.request.RawPostDataException: You cannot access body after reading from request's data stream` is thrown in the post method of my CeateAPIView whenever I try to do request.data and the error callback is executed in the ajax with a 500 internal server error.

I have no custom middleware and I'm not accessing the request in my own code (that I know of). Is anyone having a similar issue?

**Comment by EricKit** [comment]
Same issue for me. This error occurs when with a breakpoint in the post method.

**Comment by rpkilby** [comment]
Hi @Rydez and @EricKit. My guess is that your issue is separate and related to #5582. Try installing the master branch and see if your issue has been fixed.

**Comment by rivasd** [comment]
I would simply like to report that I had the same issue, and it was also caused by setting a vscode breakpoint in a GenericAPIView post() method, specifically [django-rest-auth](https://github.com/Tivix/django-rest-auth)'s LoginView. It'd be really great if this gotcha was included in some documentation, because If had not found this thread, I'd have lost even more hours debugging this problem!

**Comment by DjMoren** [comment]
Same weird vscode problem here, just to give visibility to the problem

**Comment by foobar8675** [comment]
ditto vscode issue

**Comment by carltongibson** [comment]
> `You cannot access body after reading from request's data stream`

It looks as if the VSCode debugger is accessing the `stream`, probably as part of inspecting the frame variables. If you could detect that in action (maybe by wrapping `stream` to log access) then it might be something you could report the VSCode people. 

This is out of scope for Django REST Framework. It's now implicitly documented here.

**Comment by jabelone** [comment]
Just wanted to mention what was causing this for me as this is the first result on Google. I had converted a method over to use DRF and was using `request.body`. As soon as I switched over to `request.data` (which it should be) it resolved the issue for me.

**Comment by marnanel** [comment]
I had this issue, and there was a different cause: some of the rest_framework.parser classes read the request body before your view ever sees it.

**Comment by MahadAkbar** [comment]
> I'm also getting this error and can't determine what's causing it. I'm trying to use ajax to perform a post request.
> 
> ```
> $.ajax({
>       type: 'POST',
>       url: 'http://127.0.0.1:8000/api/accounts/register/',
>       data: JSON.stringify({something: 4}),
>       cache: false,
>       contentType: 'application/json',
>       dataType: 'json',
>       success: function(data) {
>         console.log(data);
>       },
>       error: function(err) {
>         console.log(err);
>       }
>     })
>     return false;
>   });
> ```
> 
> ```
> class UserCreateAPIView(CreateAPIView):
>   permission_classes = [AllowAny]
>   serializer_class = UserCreateSerializer
> 
>   def post(self, request, *args, **kwargs):
>     serializer = UserCreateSerializer(data=request.data)
> ```
> 
> `django.http.request.RawPostDataException: You cannot access body after reading from request's data stream` is thrown in the post method of my CeateAPIView whenever I try to do request.data and the error callback is executed in the ajax with a 500 internal server error.
> 
> I have no custom middleware and I'm not accessing the request in my own code (that I know of). Is anyone having a similar issue?
> 
> UPDATE:
> I think I've resolved the problem. I'm not sure why, but the error only occurs when setting a breakpoint in the post method. Removing the breakpoint fixed the problem.

Which kind of break point you using with in post?

**Comment by Enjolras666** [comment]
Same issue，Chrome browser can't work, change other browser then it worked. Finally, I clean the Chrome data in settings making it working not just press the shortcut key<ctrl+F5>.
